monai/apps/auto3dseg/hpo_gen.py

```diff
@@ -0,0 +1,263 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from abc import abstractmethod
+from copy import deepcopy
+from typing import List
+
+from monai.apps.auto3dseg import BundleAlgo
+from monai.apps.utils import get_logger
+from monai.auto3dseg import Algo, AlgoGen, algo_export_to_pickle, algo_import_from_pickle
+from monai.bundle.config_parser import ConfigParser
+from monai.utils import optional_import
+
+nni, has_nni = optional_import("nni")
+
+logger = get_logger(module_name=__name__)
+
+__all__ = ["HPOGen", "NNIGen"]
+
+
+class HPOGen(AlgoGen):
+    """
+    This class generates a set of al, each of them can run independently.
+    """
+
+    @abstractmethod
+    def get_hyperparameters(self):
+        """Get the hyperparameter from HPO"""
+        raise NotImplementedError("")
+
+    @abstractmethod
+    def update_params(self, *args, **kwargs):  # type: ignore
+        """Update model params"""
+        raise NotImplementedError("")
+
+    @abstractmethod
+    def set_score(self):
+        """Report result to HPO"""
+        raise NotImplementedError("")
+
+    @abstractmethod
+    def run_algo(self, *args, **kwargs):  # type: ignore
+        """Interface for the HPO to run the training"""
+        raise NotImplementedError("")
+
+    @abstractmethod
+    def get_history(self, *args, **kwargs):  # type: ignore
+        """Interface for the HPO to run the history"""
+        raise NotImplementedError("")
+
+
+class NNIGen(HPOGen):
+    """
+    Generate algorithms for the NNI to automate algorithm hyper-parameter tuning. The module has two major interfaces:
+    ``__init__`` which sets up the algorithm, and a trialCommand function ``run_algo`` for the NNI library. More about
+    trialCommand function can be found in ``trail code`` section in NNI webpage
+    https://nni.readthedocs.io/en/latest/tutorials/hpo_quickstart_pytorch/main.html .
+
+    Args:
+        algo_dict: an dict that has {name: Algo object} format. The Algo object must define two methods: get_output_path
+            and train.
+        params: a set of parameter to override the algo if override is supported by Algo subclass.
+
+    Raises:
+        ValueError if the user tries to override a algo that doesn't have ``export_to_disk`` function.
+
+    Examples:
+        The experiment will keep generating new folders to save the model checkpoints, scripts, and configs if available.
+        ├── algorithm_templates
```

Comment:
this should include necessary `__init__.py`

Reply:
Yes, it seems working fine without __init__.py

Reply:
for some reason one of the model zoo items ended up using this https://github.com/Project-MONAI/model-zoo/blob/ddc97fb3b53b8a4f819855bf91c62cb033976993/models/valve_landmarks/configs/inference.json#L13
https://github.com/Project-MONAI/model-zoo/blob/ddc97fb3b53b8a4f819855bf91c62cb033976993/models/valve_landmarks/configs/inference.json#L5

adding `__init__.py` will have more flexibility but it's easy to forget to add this file...as long as we have good exception handling this should be fine.

Reply:
Added the __init__ in research contributions repo in PR and merged. Thanks! @Nic-Ma